app/src/main/java/org/thoughtcrime/securesms/conversation/v2/AttachmentDownloadHelper.kt

```diff
@@ -0,0 +1,158 @@
+
+    private val downloadRequests = Channel<DatabaseAttachment>(UNLIMITED)
+
+    /**
+     * Single use flow that buffers attachment download requests and emits them in batches. The batch
+     * will have size ([BUFFER_MAX_ITEMS]) and time([BUFFER_TIMEOUT_MILLS]) limit.
+     */
+    private val downloadRequestsFlow: Flow<List<DatabaseAttachment>> = flow {
+        var deadline = -1L
+        val buffer = mutableListOf<DatabaseAttachment>()
+
+        while (true) {
+            if (buffer.isEmpty()) {
+                // When started, wait indefinitely for the first attachment to arrive
+                buffer += downloadRequests.receive()
+                deadline = SystemClock.uptimeMillis() + BUFFER_TIMEOUT_MILLS
+                continue
+            }
+
+            if (buffer.size < BUFFER_MAX_ITEMS) {
+                // Wait for next attachment to arrive with a timeout
+                val attachment =
+                    withTimeoutOrNull((deadline - SystemClock.uptimeMillis()).coerceAtLeast(0L)) {
+                        downloadRequests.receive()
+                    }
+
+                if (attachment != null) {
+                    // Received an attachment before timeout, loop again
+                    buffer += attachment
+                    continue
+                }
+            }
+
+            // When we reach here, the buffer has reached its max size AND the timeout has expired
+            emit(buffer.toList())
+            buffer.clear()
+            deadline = -1L
+        }
+    }
+
+
+    init {
+        scope.launch {
+            downloadRequestsFlow
+                .flowOn(Dispatchers.IO)
+                .map { attachments ->
+                    val pendingAttachmentIDs = storage
+                        .getAllPendingJobs(AttachmentDownloadJob.KEY, AttachmentUploadJob.KEY)
+                        .values
+                        .mapNotNullTo(hashSetOf()) {
+                            (it as? AttachmentUploadJob)?.attachmentID
+                                ?: (it as? AttachmentDownloadJob)?.attachmentID
+                        }
+
+                    val messagesByID = mmsDatabase.getMessages(attachments.map { it.mmsId })
+                        .associateBy { it.id }
+
+                    // Before handling out attachment to the download task, we need to
+                    // check the requisite for that attachment. This check is very likely to be
+                    // performed again in the download task, but adding stuff into job system
+                    // is expensive so we need to avoid spawning new task whenever we can.
+                    attachments.filter {
+                        !pendingAttachmentIDs.contains(it.attachmentId.rowId) &&
+                                it.eligibleForDownloadTask(messagesByID[it.mmsId])
+                    }
+                }
+                .collect { attachmentsToDownload ->
+                    for (attachment in attachmentsToDownload) {
```

Comment:
I think we should generally prefer the more functional looking: `attachmentsToDownload.forEach {` as:
1. you save some parentheses
2. There's no visual-confusion between this being a for-in loop and a `Set#contains` lookup.
3. You can choose to use implcit `it` or give the var a name in the lambda.

But in this case, you could even just add some `.flatten()` or similar before `.collect {` and get around it that way.

Reply:
I actually prefer a plain loop over `.forEach` because `forEach` is a function call with an inline body, which makes debugging through it very annoying. To me it doesn't look bad cosmetically.

Reply:
Haven't found a cheap solution to flatten the list into the flow though.

Reply:
Leave however you prefer, we may talk about code conventions soon.